### ISSUE TITLE
Class weight for xgboost in mlexperiments

### ISSUE THREAD

**Issue by YC0721** [kapsner/mllrnrs#3]
Hi,
I am trying to use "weight" argument for class weight in Xgboost for. Could you please let me know how to use the class weight in in mlexperiments for grid search?

Thank you very much!

**Comment by kapsner** [comment]
Thank you for bringing this to my attention.

Unfortunately, the weight-argument (xgboost::setinfo(weight = vector)`) is currently not implemented in mlexperiments / mllrnrs.

But it could be provided, for example, when constructing the train-matrix (https://github.com/kapsner/mllrnrs/blob/main/R/learner_xgboost.R#L269). (See also https://www.rdocumentation.org/packages/xgboost/versions/1.7.7.1/topics/dim.xgb.DMatrix and https://www.rdocumentation.org/packages/xgboost/versions/1.7.7.1/topics/setinfo)

I can have a look at this and try to add this option in the next release (e.g. via mlexperiment's `learner_args` https://github.com/kapsner/mllrnrs/blob/main/vignettes/mllrnrs_xgboost_multiclass.Rmd#L107).

**Comment by kapsner** [comment]
Task list:
- [x] add feature to mllrnrs - xgboost
- [x] ensure compatibility in mlexperiments
- [x] check, if feature makes also sense for mllrnrs - lightgbm (and other algorithms) edit: YES: https://lightgbm.readthedocs.io/en/latest/R/reference/lgb.Dataset.html; https://www.rdocumentation.org/packages/ranger/versions/0.16.0/topics/ranger; https://rdrr.io/cran/glmnet/man/glmnet.html
- [x] implement feature also for lightgbm, glmnet and ranger

**Comment by kapsner** [comment]
Hi @YC0721 , I've started to implement the feature. You can try it out using the latest dev-versions of mlexperiments and mllrnrs.

```r
remotes::install_github("kapsner/mlexperiments", "development")
remotes::install_github("kapsner/mllrnrs", "development")
```

To use the feature with multiclass-tasks and `xgboost`, just add a vector of training weights (same length as your `train_y` target variable) to the `learner_args`-list under the argument `target_weights`:

```r
y_weights <- ifelse(train_y == 1, 0.8, ifelse(train_y == 2, 1.2, train_y))
validator$learner_args <- c(
  tuner$results$best.setting[-1],
  list("target_weights" = y_weights)
)
```

Then, the weights are going to be applied when creating the training dataframe. Hope, that works for you. Let me know, what you think about it!

(P.S.: I've also added examples in the vignette: https://github.com/kapsner/mllrnrs/blob/development/vignettes/mllrnrs_xgboost_multiclass.Rmd#L463)

**Comment by kapsner** [comment]
- [x] implement weights also for rprart in mlexperiments + unit tests (https://stat.ethz.ch/R-manual/R-devel/library/rpart/html/rpart.html)

**Comment by kapsner** [comment]
@YC0721 would be great to hear if you can work with the enhancements made. Please be aware that during the last commits, the `mlexperiment` argument for `learner_args` is called `case_weights` (instead of previously `target_weights`).

**Comment by YC0721** [comment]
Dear Kapsner,

Sorry for the late reply as we are still trying the your updated version
recently.

Thank you so much for your kind reminder!

Best wishes,
Yingchai

**Comment by kapsner** [comment]
Sure, just let me know if I can help anywhere.

(I've also updated the vignette's code everywhere, this could be a good starting point trying out the latest adjustments: https://github.com/kapsner/mllrnrs/blob/development/vignettes/mllrnrs_xgboost_multiclass.Rmd#L484)

**Comment by YC0721** [comment]
Dear Kapsner,

Eric (in cc) and I tried your updated version this week.

We find it should be "case_weights" and used this argument to weighted our
extremely uneven 3 classes (No of Class 1: Class 2: Class 3 = 546:325:35,
weighted with 1:1.5:15).  We find that the parameters indeed changed after
weighted, but with very minor changes compared with the unweighted one,
that we are still trying to solve it. Can we know how you set the weight in
xgboost?

Thank you for your great help!

Best wishes,
Yingchai

**Comment by kapsner** [comment]
`case_weights` is passed on to `xgboost::setinfo` to apply `xgboost`'s `weight` argument to the `dtrain` matrix: https://github.com/kapsner/mllrnrs/blob/development/R/learner_xgboost.R#L296

**Comment by YC0721** [comment]
Dear Kapsner,

Thanks for the link above.

It works well for the weight code; however, I have some problems in
variable importance after weight.


Data$weight<-NA

Data$weight[ Data$class=="Class 1"]<-1

Data$weight[ Data$class=="Class 2"]<-1.5

Data$weight[ Data$class=="Class 3"]<-15



# Grid search process (omitted)



xgbst <- xgboost(data = as.matrix(Data[, -1]),

                 label = as.numeric(Data$class) - 1,

                 weight = Data$weight,

                 max_depth = 2,

                 eta = 0.5,

                 nrounds = 16,

                 min_child_weight = 1,

                 subsample = 1.0,

                 colsample_bytree = 1,

                 nthread = 5

                )



importance_all <- xgb.importance(model = xgbst)

print (importance_all)

# Feature  Gain  Cover Frequency

 #   1               1           1           1



Could you please let me know how to get the variable importance after
weight?


Thank you very much!


Best wishes,

Yingchai

**Comment by kapsner** [comment]
Hi @YC0721 , great to hear that the implementation works for you - thus this issue will be closed now.

Regarding variable importance: `mlexperiments` / `mllrnrs` are "just" wrapper packages to provide a common code for hyperparameter optimization and validation for different learner algorithms - variable importance is not implemented there and other R packages exist for this task. Your question is very specific to `xgboost`, so I would recommend to post your question there, e.g. in the forum: https://discuss.xgboost.ai/.

Best, Lorenz
